pyproject.toml

```diff
@@ -0,0 +1,71 @@
+[project]
+name = "open-grid-emissions"
+requires-python = ">3.10"
+readme = "README.md"
+
+[tool.ruff]
+# Exclude a variety of commonly ignored directories.
+exclude = [
+    ".git",
+    ".github",
+    ".pytest_cache",
+    ".ruff_cache",
+    ".tox",
+    ".venv",
+    "__pypackages__",
+    "_build",
+    "build",
+    "venv",
+]
+
+# All paths are relative to the project root, which is the directory containing the pyproject.toml.
+src = ["src"]
+
+# Same as Black.
+line-length = 88
+indent-width = 4
+
+# Assume Python 3.8
+target-version = "py310"
+
+# Built-in for Jupyter
+extend-include = ["*.ipynb"]
+
+[tool.ruff.lint]
+# Enable Pyflakes (`F`) and a subset of the pycodestyle (`E`) codes by default.
+# Unlike Flake8, Ruff doesn't enable pycodestyle warnings (`W`) or
+# McCabe complexity (`C901`) by default.
+select = ["E4", "E7", "E9", "F"]
+ignore = []
+
+# Allow fix for all enabled rules (when `--fix`) is provided.
+fixable = ["ALL"]
+unfixable = []
+
+# Allow unused variables when underscore-prefixed.
+dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"
+
+# Exclude notebooks
+exclude = ["*.ipynb"]
```

Comment:
The notebooks are excluded from the linting because too many error are generated as shown below when I comment the `exclude` directive. Note that notebooks are formatted.
```
(open-grid-emissions) [~/Singularity/open-grid-emissions] (ben/format) brdo$ ruff check .
notebooks/explore_data/compare_data_sources.ipynb:cell 2:2:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 2:3:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 2:4:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 2:5:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 2:7:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 2:10:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 2:11:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 3:5:1: E402 Module level import not at top of file
notebooks/explore_data/compare_data_sources.ipynb:cell 3:5:29: F401 [*] `src.data_cleaning` imported but unused
notebooks/explore_data/explore_annually_reported_eia_data.ipynb:cell 1:8:1: E402 Module level import not at top of file
notebooks/explore_data/explore_annually_reported_eia_data.ipynb:cell 1:12:1: E402 Module level import not at top of file
notebooks/explore_data/explore_annually_reported_eia_data.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/explore_data/explore_annually_reported_eia_data.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/explore_data/explore_annually_reported_eia_data.ipynb:cell 1:14:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/explore_data/explore_annually_reported_eia_data.ipynb:cell 3:7:8: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 1:8:1: E402 Module level import not at top of file
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 1:12:1: E402 Module level import not at top of file
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 1:13:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 3:6:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 3:11:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 3:16:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 3:21:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 3:26:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 8:6:11: F405 `os` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 8:6:22: F405 `results_folder` may be undefined, or defined from star imports
notebooks/explore_data/explore_intermediate_outputs.ipynb:cell 8:8:9: F405 `results_folder` may be undefined, or defined from star imports
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:4:8: F401 [*] `os` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:11:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:15:8: F401 [*] `download_data` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:16:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:17:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:18:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:18:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:19:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:19:8: F401 [*] `impute_hourly_profiles` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:20:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:20:8: F401 [*] `data_cleaning` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:21:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:21:8: F401 [*] `output_data` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:22:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:22:8: F401 [*] `emissions` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:23:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:23:8: F401 [*] `validation` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:24:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:25:1: E402 Module level import not at top of file
notebooks/explore_data/gens_not_in_cems.ipynb:cell 1:25:8: F401 [*] `eia930` imported but unused
notebooks/explore_data/gens_not_in_cems.ipynb:cell 2:3:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/gens_not_in_cems.ipynb:cell 2:8:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/gens_not_in_cems.ipynb:cell 2:13:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/gens_not_in_cems.ipynb:cell 2:18:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/gens_not_in_cems.ipynb:cell 2:23:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/explore_data/manually_import_uncontrolled_emission_factors.ipynb:cell 2:6:1: E402 Module level import not at top of file
notebooks/explore_data/manually_import_uncontrolled_emission_factors.ipynb:cell 2:6:29: F401 [*] `src.data_cleaning` imported but unused
notebooks/explore_data/manually_import_uncontrolled_emission_factors.ipynb:cell 2:7:1: E402 Module level import not at top of file
notebooks/explore_data/manually_import_uncontrolled_emission_factors.ipynb:cell 2:7:25: F401 [*] `src.load_data` imported but unused
notebooks/explore_data/manually_import_uncontrolled_emission_factors.ipynb:cell 2:8:1: E402 Module level import not at top of file
notebooks/explore_data/manually_import_uncontrolled_emission_factors.ipynb:cell 7:31:5: F841 Local variable `map_eia_combustion_system_to_ours` is assigned to but never used
notebooks/explore_data/manually_import_uncontrolled_emission_factors.ipynb:cell 7:77:9: F841 Local variable `mapped_units_dict` is assigned to but never used
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 3:1:17: F401 [*] `numpy` imported but unused
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 3:5:32: F401 [*] `plotly.graph_objects` imported but unused
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 3:7:20: F401 [*] `datetime` imported but unused
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:6:1: E402 Module level import not at top of file
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:10:1: E402 Module level import not at top of file
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:10:31: F401 [*] `src.visualization.day_hour_heatmap` imported but unused
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:11:1: E402 Module level import not at top of file
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:12:5: F401 [*] `src.eia930.fuel_code_map` imported but unused
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:13:5: F401 [*] `src.eia930.reformat_chalendar` imported but unused
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:14:5: F401 [*] `src.eia930.load_chalendar` imported but unused
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:17:1: E402 Module level import not at top of file
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:18:1: E402 Module level import not at top of file
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 4:19:1: E402 Module level import not at top of file
notebooks/explore_methods/calculate_residual_net_generation.ipynb:cell 17:50:39: E712 Comparison to `False` should be `cond is False` or `if not cond:`
notebooks/explore_methods/national_wind_solar_correlations.ipynb:cell 2:2:18: F401 [*] `pandas` imported but unused
notebooks/explore_methods/national_wind_solar_correlations.ipynb:cell 2:8:1: E402 Module level import not at top of file
notebooks/explore_methods/national_wind_solar_correlations.ipynb:cell 2:13:1: E402 Module level import not at top of file
notebooks/explore_methods/national_wind_solar_correlations.ipynb:cell 2:15:1: E402 Module level import not at top of file
notebooks/explore_methods/national_wind_solar_correlations.ipynb:cell 2:15:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/explore_methods/national_wind_solar_correlations.ipynb:cell 3:4:21: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/manual_data/default_fuel_sulfur_content.ipynb:cell 2:3:18: F401 [*] `pandas` imported but unused
notebooks/manual_data/default_fuel_sulfur_content.ipynb:cell 2:4:26: F401 [*] `plotly.express` imported but unused
notebooks/manual_data/default_fuel_sulfur_content.ipynb:cell 2:7:1: E402 Module level import not at top of file
notebooks/manual_data/default_fuel_sulfur_content.ipynb:cell 2:8:1: E402 Module level import not at top of file
notebooks/manual_data/export_fuel_heat_content.ipynb:cell 1:3:26: F401 [*] `plotly.express` imported but unused
notebooks/manual_data/export_fuel_heat_content.ipynb:cell 1:9:1: E402 Module level import not at top of file
notebooks/manual_data/export_fuel_heat_content.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/manual_data/export_fuel_heat_content.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/manual_data/export_fuel_heat_content.ipynb:cell 1:14:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/manual_data/export_fuel_heat_content.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/manual_data/export_fuel_heat_content.ipynb:cell 3:20:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 2:2:17: F401 [*] `numpy` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 2:4:22: F401 [*] `datetime.timedelta` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:5:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:9:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:9:8: F401 [*] `download_data` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:10:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:10:8: F401 [*] `load_data` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:11:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:12:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:12:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:13:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:13:8: F401 [*] `impute_hourly_profiles` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:14:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:14:8: F401 [*] `data_cleaning` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:15:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:15:8: F401 [*] `output_data` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:16:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:16:8: F401 [*] `emissions` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:17:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:17:8: F401 [*] `validation` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:18:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:18:8: F401 [*] `gross_to_net_generation` imported but unused
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 3:19:1: E402 Module level import not at top of file
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 8:6:15: F405 `data_folder` may be undefined, or defined from star imports
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 9:2:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/manual_data/identify_eia930_time_lags.ipynb:cell 36:1:11: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/manual_data/manually_update_OTH_fuel_code.ipynb:cell 1:5:1: E402 Module level import not at top of file
notebooks/manual_data/manually_update_OTH_fuel_code.ipynb:cell 1:6:1: E402 Module level import not at top of file
notebooks/manual_data/manually_update_OTH_fuel_code.ipynb:cell 1:6:18: F401 [*] `pandas` imported but unused
notebooks/manual_data/manually_update_OTH_fuel_code.ipynb:cell 1:7:1: E402 Module level import not at top of file
notebooks/manual_data/manually_update_OTH_fuel_code.ipynb:cell 1:9:1: E402 Module level import not at top of file
notebooks/manual_data/manually_update_OTH_fuel_code.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/manual_data/manually_update_OTH_fuel_code.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/manual_data/manually_update_ba_reference.ipynb:cell 2:1:8: F401 [*] `os` imported but unused
notebooks/manual_data/manually_update_ba_reference.ipynb:cell 2:2:8: F401 [*] `requests` imported but unused
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:4:8: F401 [*] `os` imported but unused
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:5:26: F401 [*] `plotly.express` imported but unused
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:11:1: E402 Module level import not at top of file
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:16:1: E402 Module level import not at top of file
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:16:27: F401 [*] `column_checks.get_dtypes` imported but unused
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:17:1: E402 Module level import not at top of file
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 1:17:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/manual_data/update_utility_name_ba_map.ipynb:cell 2:25:34: F405 `manual_folder` may be undefined, or defined from star imports
notebooks/manual_data/zip_data.ipynb:cell 2:5:1: E402 Module level import not at top of file
notebooks/manual_data/zip_data.ipynb:cell 2:9:1: E402 Module level import not at top of file
notebooks/validation/data_validation.ipynb:cell 1:4:26: F401 [*] `plotly.express` imported but unused
notebooks/validation/data_validation.ipynb:cell 2:5:1: E402 Module level import not at top of file
notebooks/validation/data_validation.ipynb:cell 2:10:1: E402 Module level import not at top of file
notebooks/validation/data_validation.ipynb:cell 2:11:1: E402 Module level import not at top of file
notebooks/validation/data_validation.ipynb:cell 2:12:1: E402 Module level import not at top of file
notebooks/validation/data_validation.ipynb:cell 2:14:1: E402 Module level import not at top of file
notebooks/validation/data_validation.ipynb:cell 2:14:39: F401 [*] `column_checks.apply_dtypes` imported but unused
notebooks/validation/diff_output_versions.ipynb:cell 1:10:1: E402 Module level import not at top of file
notebooks/validation/diff_output_versions.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/validation/diff_output_versions.ipynb:cell 1:14:8: F401 [*] `load_data` imported but unused
notebooks/validation/diff_output_versions.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/validation/diff_output_versions.ipynb:cell 1:16:1: E402 Module level import not at top of file
notebooks/validation/diff_output_versions.ipynb:cell 1:16:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/validation/diff_output_versions.ipynb:cell 4:6:23: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 4:7:14: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 4:9:5: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 4:12:9: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 4:17:5: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 4:23:5: F405 `results_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 4:29:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:6:23: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:7:14: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:9:5: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:12:9: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:17:22: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:19:9: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:29:22: F405 `results_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 7:31:9: F405 `results_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 9:6:23: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 9:7:14: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 9:9:9: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 9:11:33: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 9:15:5: F405 `data_folder` may be undefined, or defined from star imports
notebooks/validation/diff_output_versions.ipynb:cell 9:19:24: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/validation/hourly_validation.ipynb:cell 2:8:22: F401 [*] `datetime.datetime` imported but unused
notebooks/validation/hourly_validation.ipynb:cell 2:10:22: F401 [*] `datetime.timedelta` imported but unused
notebooks/validation/hourly_validation.ipynb:cell 3:5:1: E402 Module level import not at top of file
notebooks/validation/hourly_validation.ipynb:cell 12:5:7: F541 [*] f-string without any placeholders
notebooks/validation/validate_data_cleaning.ipynb:cell 1:8:1: E402 Module level import not at top of file
notebooks/validation/validate_data_cleaning.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/validation/validate_data_cleaning.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:8:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:8:43: F401 [*] `src.column_checks.apply_dtypes` imported but unused
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:9:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:10:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:11:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:11:25: F401 [*] `src.load_data` imported but unused
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:12:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/validation/validate_hourly_profiles.ipynb:cell 1:14:17: F401 [*] `numpy` imported but unused
notebooks/validation/validate_negative_profiles.ipynb:cell 1:10:1: E402 Module level import not at top of file
notebooks/validation/validate_negative_profiles.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/validation/validate_negative_profiles.ipynb:cell 1:15:25: F401 [*] `src.load_data` imported but unused
notebooks/validation/validate_negative_profiles.ipynb:cell 1:16:1: E402 Module level import not at top of file
notebooks/validation/validate_negative_profiles.ipynb:cell 1:16:29: F401 [*] `src.data_cleaning` imported but unused
notebooks/validation/validate_negative_profiles.ipynb:cell 1:17:1: E402 Module level import not at top of file
notebooks/validation/validate_negative_profiles.ipynb:cell 1:18:1: E402 Module level import not at top of file
notebooks/validation/validate_negative_profiles.ipynb:cell 1:18:38: F401 [*] `src.impute_hourly_profiles` imported but unused
notebooks/validation/validate_negative_profiles.ipynb:cell 1:21:1: E402 Module level import not at top of file
notebooks/validation/validate_negative_profiles.ipynb:cell 1:21:43: F401 [*] `src.column_checks.apply_dtypes` imported but unused
notebooks/validation/validate_vs_egrid.ipynb:cell 1:2:17: F401 [*] `numpy` imported but unused
notebooks/validation/validate_vs_egrid.ipynb:cell 1:5:26: F401 [*] `plotly.express` imported but unused
notebooks/validation/validate_vs_egrid.ipynb:cell 1:12:1: E402 Module level import not at top of file
notebooks/validation/validate_vs_egrid.ipynb:cell 1:17:1: E402 Module level import not at top of file
notebooks/validation/validate_vs_egrid.ipynb:cell 1:18:1: E402 Module level import not at top of file
notebooks/validation/validate_vs_egrid.ipynb:cell 1:19:1: E402 Module level import not at top of file
notebooks/validation/validate_vs_egrid.ipynb:cell 1:19:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/validation/validate_vs_egrid.ipynb:cell 1:21:1: E402 Module level import not at top of file
notebooks/validation/validate_vs_egrid.ipynb:cell 5:3:5: F405 `results_folder` may be undefined, or defined from star imports
notebooks/validation/validate_vs_egrid.ipynb:cell 5:7:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/validation/validate_vs_egrid.ipynb:cell 22:3:5: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/validation/validate_vs_egrid.ipynb:cell 33:5:5: F405 `results_folder` may be undefined, or defined from star imports
notebooks/validation/validate_vs_egrid.ipynb:cell 33:9:9: F405 `results_folder` may be undefined, or defined from star imports
notebooks/visualization/map_visualization.ipynb:cell 2:3:1: F403 `from plotly.colors import *` used; unable to detect undefined names
notebooks/visualization/map_visualization.ipynb:cell 2:4:21: F401 [*] `plotly.io` imported but unused
notebooks/visualization/map_visualization.ipynb:cell 3:5:1: E402 Module level import not at top of file
notebooks/visualization/map_visualization.ipynb:cell 3:9:1: E402 Module level import not at top of file
notebooks/visualization/map_visualization.ipynb:cell 3:9:8: F401 [*] `output_data` imported but unused
notebooks/visualization/map_visualization.ipynb:cell 9:2:1: E402 Module level import not at top of file
notebooks/visualization/map_visualization.ipynb:cell 11:20:18: F405 `diverging` may be undefined, or defined from star imports
notebooks/visualization/map_visualization.ipynb:cell 15:9:14: F405 `diverging` may be undefined, or defined from star imports
notebooks/visualization/map_visualization.ipynb:cell 15:190:17: F541 [*] f-string without any placeholders
notebooks/visualization/plot_timeseries_data.ipynb:cell 1:9:1: E402 Module level import not at top of file
notebooks/visualization/plot_timeseries_data.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/visualization/plot_timeseries_data.ipynb:cell 1:13:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/visualization/plot_timeseries_data.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/visualization/plot_timeseries_data.ipynb:cell 1:14:8: F401 [*] `validation` imported but unused
notebooks/visualization/plot_timeseries_data.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/visualization/plot_timeseries_data.ipynb:cell 8:5:5: F405 `results_folder` may be undefined, or defined from star imports
notebooks/work_in_progress/GH102_test_dask.ipynb:cell 2:2:17: F401 [*] `numpy` imported but unused
notebooks/work_in_progress/GH102_test_dask.ipynb:cell 4:5:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH102_test_dask.ipynb:cell 4:10:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:8:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:11:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:11:27: F401 [*] `column_checks.get_dtypes` imported but unused
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:12:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:13:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:13:1: F403 `from data_cleaning import *` used; unable to detect undefined names
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:14:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:14:1: F403 `from emissions import *` used; unable to detect undefined names
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:16:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 1:16:44: F401 [*] `pudl.analysis.allocate_net_gen.distribute_annually_reported_data_to_months` imported but unused
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 4:6:8: F405 `remove_plants` may be undefined, or defined from star imports
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 4:15:8: F405 `manually_remove_steam_units` may be undefined, or defined from star imports
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 4:22:8: F405 `remove_incomplete_unit_months` may be undefined, or defined from star imports
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 4:27:8: F405 `assign_fuel_type_to_cems` may be undefined, or defined from star imports
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 8:1:29: F405 `emissions` may be undefined, or defined from star imports
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 11:2:22: F405 `emissions` may be undefined, or defined from star imports
notebooks/work_in_progress/GH153_fill_missing_nox_so2_cems.ipynb:cell 11:3:39: F405 `pudl_out` may be undefined, or defined from star imports
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:2:18: F401 [*] `pandas` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:3:8: F401 [*] `os` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:10:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:14:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:14:8: F401 [*] `download_data` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:15:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:15:8: F401 [*] `load_data` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:16:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:16:27: F401 [*] `column_checks.get_dtypes` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:17:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:17:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:18:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:18:8: F401 [*] `impute_hourly_profiles` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:19:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:19:8: F401 [*] `data_cleaning` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:20:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:20:8: F401 [*] `output_data` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:21:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:21:8: F401 [*] `emissions` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:22:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:22:8: F401 [*] `validation` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:23:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:23:8: F401 [*] `gross_to_net_generation` imported but unused
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 2:24:1: E402 Module level import not at top of file
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 3:2:16: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/work_in_progress/GH240_eia930_physics_reconciliation.ipynb:cell 3:3:18: F405 `outputs_folder` may be undefined, or defined from star imports
notebooks/work_in_progress/clean_cems_outliers.ipynb:cell 3:5:1: E402 Module level import not at top of file
notebooks/work_in_progress/clean_cems_outliers.ipynb:cell 3:10:1: E402 Module level import not at top of file
notebooks/work_in_progress/clean_cems_outliers.ipynb:cell 3:11:1: E402 Module level import not at top of file
notebooks/work_in_progress/clean_cems_outliers.ipynb:cell 3:12:1: E402 Module level import not at top of file
notebooks/work_in_progress/issue230_spikes.ipynb:cell 3:5:1: E402 Module level import not at top of file
notebooks/work_in_progress/issue230_spikes.ipynb:cell 3:9:1: E402 Module level import not at top of file
notebooks/work_in_progress/issue230_spikes.ipynb:cell 3:9:8: F401 [*] `impute_hourly_profiles` imported but unused
notebooks/work_in_progress/issue230_spikes.ipynb:cell 3:10:1: E402 Module level import not at top of file
notebooks/work_in_progress/issue230_spikes.ipynb:cell 3:11:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:2:18: F401 [*] `pandas` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:3:17: F401 [*] `numpy` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:4:8: F401 [*] `os` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:5:26: F401 [*] `plotly.express` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:11:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:15:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:15:8: F401 [*] `download_data` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:16:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:16:8: F401 [*] `load_data` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:17:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:17:27: F401 [*] `column_checks.get_dtypes` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:18:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:18:1: F403 `from filepaths import *` used; unable to detect undefined names
notebooks/work_in_progress/sandbox.ipynb:cell 1:19:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:19:8: F401 [*] `impute_hourly_profiles` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:20:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:20:8: F401 [*] `data_cleaning` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:21:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:21:8: F401 [*] `output_data` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:22:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:22:8: F401 [*] `emissions` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:23:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:23:8: F401 [*] `validation` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:24:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:24:8: F401 [*] `gross_to_net_generation` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:25:1: E402 Module level import not at top of file
notebooks/work_in_progress/sandbox.ipynb:cell 1:25:8: F401 [*] `eia930` imported but unused
notebooks/work_in_progress/sandbox.ipynb:cell 1:26:1: E402 Module level import not at top of file
notebooks/work_in_progress/uncertainty_analysis.ipynb:cell 3:6:1: E402 Module level import not at top of file
notebooks/work_in_progress/uncertainty_analysis.ipynb:cell 3:10:1: E402 Module level import not at top of file
notebooks/work_in_progress/uncertainty_analysis.ipynb:cell 3:11:1: E402 Module level import not at top of file
notebooks/work_in_progress/uncertainty_analysis.ipynb:cell 3:11:29: F401 [*] `src.column_checks` imported but unused
Found 315 errors.
[*] 87 fixable with the `--fix` option (3 hidden fixes can be enabled with the `--unsafe-fixes` option).
```
 only 87/315 can be fixed automatically. We would need to tackle the remainder by hand

Reply:
Haha ok let's keep ignoring them for now. We should at some point probably clean up the notebooks anyway.